src/testers.jl

```diff
@@ -285,3 +269,58 @@ function _is_inferrable(f, args...; kwargs...)
         return false
     end
 end
+
+"""
+    _test_cotangent(accum_cotangent, ad_cotangent, fd_cotangent; kwargs...)
+
+Check if the cotangent `ad_cotangent` from `rrule` is consistent with `accum_tangent` and
+approximately equal to the cotangent `fd_cotangent` obtained with finite differencing.
+
+If `accum_cotangent` is `NoTangent()`, i.e., the argument was marked as non-differentiable,
+`ad_cotangent` and `fd_cotangent` should be `NoTangent()` as well.
+
+# Keyword arguments
+- If `check_inferred=true` (the default) and `ad_cotangent` is a thunk, then it is checked if
+  its content can be inferred.
+- All remaining keyword arguments are passed to `isapprox`.
+"""
+function _test_cotangent(
+    accum_cotangent,
+    ad_cotangent,
+    fd_cotangent;
+    check_inferred=true,
+    kwargs...,
+)
+    ad_cotangent isa AbstractThunk && check_inferred && _test_inferred(unthunk, ad_cotangent)
+
+    # The main test of the actual derivative being correct:
+    test_approx(ad_cotangent, fd_cotangent; kwargs...)
+    _test_add!!_behaviour(accum_cotangent, ad_cotangent; kwargs...)
+end
+
+# we marked the argument as non-differentiable
+function _test_cotangent(::NoTangent, ad_cotangent, ::NoTangent; kwargs...)
+    @test ad_cotangent isa NoTangent
+end
+function _test_cotangent(::NoTangent, ::ZeroTangent, ::NoTangent; kwargs...)
+    error(
+        "The pullback in the rrule should use NoTangent()" *
+        "rather than ZeroTangent() for non-perturbable arguments."
+    )
+end
+function _test_cotangent(
+    ::NoTangent,
+    ad_cotangent::ChainRulesCore.NotImplemented,
+    ::NoTangent;
+    kwargs...,
+)
+    # this situation can occur if a cotangent is not implemented and
+    # the default `rand_tangent` is `NoTangent`: e.g. due to having no fields
+    # the `@test_broken` below should tell them that there is an easy implementation for
+    # this case of `NoTangent()` (`@test_broken false` would be less useful!)
+    # https://github.com/JuliaDiff/ChainRulesTestUtils.jl/issues/217
+    @test_broken ad_cotangent isa NoTangent
```

Comment:
> the `@test_broken` below should tell them that there is an easy implementation for this case of `NoTangent()`

I don't understand what this is supposed to say, could you explain so we can make it a more helpful comment for future readers of the codebase?

Reply:
From the comment by @oxinabox on the outdated diff:

> The correct implementation of anything that test_rrule determined was a NoTangent is almost certainly NoTangent()
Which is easy.
It might not generalize to cases that are not tested, but for the types passed into this test it is (almost certainly) correct.

But I'm still not clear what I'd be expected to do.

Reply:
If the argument is actually differentiable and `NoTangent` is just caused by an incorrect `rand_tangent` default, then you should provide a correct tangent.

Reply:
Where would we have ended up if `fun` _had_ fields? How would I tell the dispatcher that an rrule would only apply to functions with no field?

Reply:
If my example
```julia
function ChainRulesCore.rrule(::typeof(foo), K, fun)
    f = foo(K, fun)

    function pullback(Δf)
        ∂self = NoTangent()
        ∂K = @thunk(2Δf)
        ∂fun = NoTangent()
        return (∂self, ∂K, ∂fun)
    end

    return f, pullback
end
```
should be generic for `fun` (the derivative for fields of `fun` _would_ exist, I just haven't worked it out yet), is `@not_implemented` not the right thing to do ?

Reply:
In general, it is not the right thing to do, the right thing is to implement the derivatives :stuck_out_tongue:  `@not_implemented` is a workaround until you've done this, and therefore all tests with `NotImplemented` use `@test_broken` to indicate that something is broken and should be fixed.